File: iosApp/CheckoutViewModel.swift
import Foundation
import shared

/// UI state for the checkout screen.
enum CheckoutUiState {
    case idle
    case loading
    case verifying
    case success
    case error(String)
}

/// ViewModel that drives the checkout SwiftUI View.
///
/// Programs against PaymentSdk — the same code works
/// for both StorePaymentSdk and InHousePaymentSdk.
///
/// Flow:
/// 1. loadProducts()    -> getProducts()
/// 2. onBuyClicked()    -> purchase() -> postReceipt()
///    (the SDK acknowledges / finishes the transaction
///    internally inside postReceipt)
@MainActor
class CheckoutViewModel: ObservableObject {

    private let paymentSdk: PaymentSdk

    @Published var products: [Product] = []
    @Published var uiState: CheckoutUiState = .idle

    init(paymentSdk: PaymentSdk) {
        self.paymentSdk = paymentSdk
    }

    // MARK: - Load Products

    func loadProducts() async {
        uiState = .loading

        products = (try? await paymentSdk.getProducts(
            productIds: ["premium_monthly", "premium_yearly"]
        )) ?? []

        uiState = .idle
    }

    // MARK: - Purchase

    func onBuyClicked(product: Product) {
        Task {
            uiState = .loading

            guard let result = try? await
                paymentSdk.purchase(product: product)
            else {
                uiState = .error("Purchase failed")
                return
            }

            switch result {

            case let success as PurchaseResult.Success:
                uiState = .verifying

                let tx = try? await
                    paymentSdk.postReceipt(purchase: success)

                if tx?.status == .completed ||
                   tx?.status == .verified {
                    uiState = .success
                } else {
                    uiState = .error("Transaction failed")
                }

            case is PurchaseResult.UserCanceled:
                uiState = .idle

            case let error as PurchaseResult.Error:
                uiState = .error(error.message)

            default:
                uiState = .idle
            }
        }
    }
}

File: androidApp/src/main/kotlin/com/example/app/CheckoutViewModel.kt
package com.example.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.paymentsdk.PaymentSdk
import com.example.paymentsdk.models.Product
import com.example.paymentsdk.models.PurchaseResult
import com.example.paymentsdk.models.TransactionStatus
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface CheckoutUiState {
    data object Idle : CheckoutUiState
    data object Loading : CheckoutUiState
    data object Verifying : CheckoutUiState
    data object Success : CheckoutUiState
    data class Error(val message: String) : CheckoutUiState
}

/**
 * ViewModel that drives the checkout Composable.
 *
 * Programs against [PaymentSdk] — the same code works
 * for both StorePaymentSdk and InHousePaymentSdk.
 *
 * Flow:
 * 1. loadProducts()    -> getProducts()
 * 2. onBuyClicked()    -> purchase() -> postReceipt()
 *    (the SDK acknowledges / finishes the transaction
 *    internally inside postReceipt)
 */
class CheckoutViewModel(
    private val paymentSdk: PaymentSdk
) : ViewModel() {

    private val _products =
        MutableStateFlow<List<Product>>(emptyList())
    val products = _products.asStateFlow()

    private val _uiState =
        MutableStateFlow<CheckoutUiState>(CheckoutUiState.Idle)
    val uiState = _uiState.asStateFlow()

    fun loadProducts(productIds: List<String>) {
        viewModelScope.launch {
            _uiState.value = CheckoutUiState.Loading
            _products.value =
                paymentSdk.getProducts(productIds)
            _uiState.value = CheckoutUiState.Idle
        }
    }

    fun onBuyClicked(product: Product) {
        viewModelScope.launch {
            _uiState.value = CheckoutUiState.Loading

            when (val result = paymentSdk.purchase(product)) {

                is PurchaseResult.Success -> {
                    _uiState.value = CheckoutUiState.Verifying

                    val tx = paymentSdk.postReceipt(result)

                    if (tx.status == TransactionStatus.COMPLETED ||
                        tx.status == TransactionStatus.VERIFIED
                    ) {
                        _uiState.value = CheckoutUiState.Success
                    } else {
                        _uiState.value = CheckoutUiState.Error(
                            "Transaction failed"
                        )
                    }
                }

                is PurchaseResult.UserCanceled -> {
                    _uiState.value = CheckoutUiState.Idle
                }

                is PurchaseResult.Error -> {
                    _uiState.value =
                        CheckoutUiState.Error(result.message)
                }
            }
        }
    }
}
